blackbird_python/blackbird/listener.py

```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
             self._program._operations.append(operation)
 
     def enterForloop(self, ctx: blackbirdParser.ForloopContext):
-        self._infor = True
+        self._in_for = True
 
     def exitForloop(self, ctx: blackbirdParser.ForloopContext):
-        self._infor = False
+        self._in_for = False
 
-        if ctx.INT():
-            _FORVAR = range(int(ctx.INT().getText()))
+        if ctx.rangeval():
+            for_var = range(*[
+                int(c.getText()) for c in ctx.rangeval().getChildren()
+                if c.getText() != ":"
+            ])
         elif ctx.vallist():
-            _FORVAR = ctx.vallist().getText().split(',')
+            for_var = [
+                _expression(c.expression()) for c in ctx.vallist().getChildren()
+                if isinstance(c, blackbirdParser.ValContext)
+            ]
 
-        for var in _FORVAR:
+        for var in for_var:
             if ctx.NAME():
                 try:
-                    var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    new_var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    if new_var != var:
+                        raise ValueError
```

Comment:
How come this `ValueError` doesn't contain a message?

I have to admit, I'm having a bit of trouble following this `try`-`except`. It could benefit from line comments!

Reply:
This `ValueError` will be excepted by the line below, thus never actually printed to the user. There might be a nicer way to do this (it's a bit hacky I guess). :thinking: 

So, basically, it tests to see if the variable is able to be casted to the defined type and, if so, it checks whether the casted variable has the same value as the old one (e.g. `5 != int(5.0)`) and if not, it raises the `invalid value` `ValueError`)